### DISCUSSION TITLE
Observable not updating

### DISCUSSION THREAD

**Discussion by PoojaC94** [microsoft/fast#6392] [General]
I have a FAST component which has multiple components in it
Component A
`const template = html<item>
${when x=> x.isComponentOpen,
<componentB></componentB>
}`
`<componentC :onClick={(x)=>x.onClick}></componentC>`

`@observable isComponentOpen = false;`
`onClick() {
this.isComponentOpen=true
}`

When componentC calls onClick, the isComponentOpen is set to true but componentB is not appearing. Why does this happen?

**Comment by KingOfTac** [comment]
@PoojaC94 Can you provide some environment information for me?
* FAST packages you are using with version numbers
* Typescript version (if applicable)
* tsconfig (if using typescript)
* Node version

I tried to repro the issue but everything seems to be working fine on my end. https://stackblitz.com/edit/typescript-warbli?file=index.ts

A couple of things to watch out for:
1. If using typescript. If your tsconfig `target` is set to "ES2022" or "ESNext" make sure that `useDefineForClassFields` is false or else decorators and dependency injection won't work properly.
2. If you're not using typescript then decorators won't work at all and the manual observable approach is needed. https://www.fast.design/docs/fast-element/observables-and-state#access-tracking
3. The second parameter of a `when` directive needs to be a `ViewTemplate` instance. I can't tell if your example is the actual code you tried or pseudo code, but an example of using `when`  in this case would be:
```ts
const template = html`${when(x => x.isComponentOpen, html`<component-b></component-b>`)}
```
4. I highly recommend reading through the [building components](https://www.fast.design/docs/category/building-components) section of the docs and building something as you follow along.

**Reply by PoojaC94** [reply]
node version is 14.18.3
typescript is 4.6.4
fast-element 1.7.0
fast-foundation 2.32.0
fast-web-utilities 5.0.0

**Reply by PoojaC94** [reply]
I had followed 3 to build a viewTemplate instance. I typed a psuedocode

**Reply by KingOfTac** [reply]
Can you add your ts-config as well?

**Reply by PoojaC94** [reply]
My tsconfig has references to other files so I can share the skeleton
"module": "esnext",
    "target": "es2016",

**Reply by KingOfTac** [reply]
Can you try changing the `:onClick` binding on `ComponentC` to `<component-c @click="${x => x.onClick()}"></component-c>`?

**Reply by PoojaC94** [reply]
unfortunately, the params in onClick are dependent on child components so it's difficult to feed it directly from parent

**Reply by PoojaC94** [reply]
componentA has a list of menuItems of which componentC is a part. and componentC has a list of menu items which, on clicking, calls onClick() in componentA

**Reply by KingOfTac** [reply]
Can you post a stackblitz repro or some code examples of what you're trying to accomplish? I'm not sure I follow what you mean. From what I understand, you want `ComponentB` to appear when `ComponentC` is clicked, in your pseudo code above, this won't work because you're trying to bind your click handler with a property binding (`:onClick`) so you would need to change that to an _event_ binding (`@click`).

**Reply by PoojaC94** [reply]
yes. your understanding is correct. how do I change it into event binding? from :click to @click?

**Reply by PoojaC94** [reply]
but componentC has a list of menuitems which use  @click="${(x, c) =>c.parent.onClick(x, c.event)

**Reply by KingOfTac** [reply]
```ts
const template = html`
  ${when(x => x.isComponentOpen, html`<component-b></component-b>`)}
  <component-c @click="${x => x.onClick()}"></component-c>
';
```

**Reply by KingOfTac** [reply]
My stackblitz example linked above has functioning code. https://stackblitz.com/edit/typescript-warbli?file=index.ts

**Reply by PoojaC94** [reply]
when componentC has multiple elements that use onClick, will that work?

**Reply by KingOfTac** [reply]
> but componentC has a list of menuitems which use @click="${(x, c) =>c.parent.onClick(x, c.event)

So `ComponentC`'s click handler is getting called from it's child elements?

**Reply by PoojaC94** [reply]
yes. the flow is, click componentC->a list of child elements appear->click on one of them->onClick() is called-> check if one condition is satisfied->set isComponentOpen to true->componentB appears

**Reply by KingOfTac** [reply]
Ah, I see. Give me a minute to try something out.

**Reply by KingOfTac** [reply]
@PoojaC94 Ok, I updated my stackblitz https://stackblitz.com/edit/typescript-warbli?file=index.ts.

Let's break down what is going on here step by step through each component. We'll start from Component C and work our way up to A.

### **Component C**
Component C essentially has two parts; a click target to show/hide its menu items and the menu items.
When we click on the title of C it will set C's `menuOpen` property which controls the visibility of the menu.
When one of the menu items gets clicked, C will emit a custom event called `component-c-item-click` (more on this later).

### **Component B**
Component B's visibility is controlled by A and doesn't really do much of anything.

### **Component A**
Component A is basically just a container around B and C to manage some state between them so it really only needs to respond to events.

Component A has a `isComponentOpen` property that controls the visibility of B.
Component A also has an event binding on C that is listening for the `component-c-item-click` event that gets emitted from C.
So now when that event gets emitted, A's `toggleComponentB` method gets called which sets the `isComponentOpen` property.


So the flow is now:
`click C => show C items` => `click C item => emit custom event from C` => `A listening to C emitting event => toggle B visibility`

**Reply by PoojaC94** [reply]
thank you! this is very helpful

**Reply by PoojaC94** [reply]
one more question, if I need to use the items in C as a component in A, how do you ensure that binding?

**Reply by KingOfTac** [reply]
Working with nested components like this where they need to interact and share state can be tricky so custom events are your friend here as they greatly simplify things over trying to have the components interact through directly mutating properties on each other.

This way the components are loosely coupled and don't develop a hard dependency on each other. Which would allow you to swap out `ComponentC` for other components based on different needs and `ComponentA` won't care because it is only looking for an event and not a specific component.

**Reply by PoojaC94** [reply]
There is `items` in componentC. After checking some of the properties of C's components, componentA triggers the isComponentOpen property

**Reply by PoojaC94** [reply]
is that possible?

**Reply by PoojaC94** [reply]
yes

**Reply by PoojaC94** [reply]
the components are created in C but accessed in A

**Reply by KingOfTac** [reply]
Updated the stackblitz. https://stackblitz.com/edit/typescript-warbli?file=index.ts

I added a data attribute to the menu items called `data-thing`. When C emits the `component-c-item-click` event, it passes the value of that data attribute into the event detail. So when A's `toggleComponentB` gets called, it can look at the value of `event.detail` to determine if that value meets the criteria for showing B.

You could create a dedicated component for the Items if you plan on using the pattern in other places, but the idea is you pass the things you need to check in A, through the event detail so that A doesn't need to know the implementation details of the other components. All it needs to care about is the event.

So in my example the first and last menu items will toggle B, but not the second item because the value of `data-thing` does not meet the criteria that A is looking for.

**Reply by PoojaC94** [reply]
so, C contains ` ${repeat(
        () => SAMPLE_ITEMS,
        html<SampleListItem>`
          <a
            class="nav-link ellipsis dropdown-item"
            title="${(x) => x.Descr}"
            @click="${(x, c) => c.parent.onClick(x, c.event)}"
          >
            ${(x) => x.Text}
          </a>
        `
      )}`
as a part of its template
when onClick is called in ComponentA, we need to analyze SampleListItem and then trigger isComponentOpen. (SampleListItem has other components that will be analyzed in onClick so it's tough to only get a single component. Will the above event.detail cover an object instead of a string?

**Reply by KingOfTac** [reply]
Yes, you can put any value into an event detail. What you'll want to do in this case is create an interface that either `SampleListItem` implements or one of its properties and then type the event detail to that interface.
```ts
interface ComplexType {
  foo: boolean;
  bar: string;
  baz: Array<number>
}

interface SomeCommonInterface {
  propertyA: boolean;
  propertyB: string;
  propertyC: ComplexType;
}

class SampleListItem implements SomeCommonInterface {}

// In Component A
// CustomEvent can take a type param so that you can set the type of detail
const template = html`
  <component-c
    @component-c-item-click="${(x, c) => x.toggleComponentB(c.event as CustomEvent<SomeCommonInterface>)}"
  ></component-c>
`;
// Now event.detail in toggleComponentB will be typed as SomeCommonInterface
```

In `C`'s event emitter you could even pass in the entire event.target if that is what you want. `A` still won't need to know the exact details of the data passed into event.detail because it only cares that it fits the shape of SomeCommonInterface. So you could pass in another component, an object literal, or even a function expression as long as the resulting data structure is what `A` is expecting.

**Reply by KingOfTac** [reply]
What is essentially happening here is instead of saying `ComponentA` requires `SampleListItem` to have x and y properties, you're forming a contract that says `ComponentA` requires something that meets the criteria of `SomeCommonInterface`. So anything can provide that to `ComponentA` as long as it agrees to that contract

**Reply by KingOfTac** [reply]
In vNext, this pattern will be even simpler with the use of the new `Context` API

**Reply by PoojaC94** [reply]
so, in x.toggleComponentB, can I do
`
toggleComponentB(event: CustomEvent<SomeCommonInterface>)`
`{
if(event.detail.propertyB == "sample"){
this.isComponentBOpen = !this.isComponentBOpen ;
}
}
`

**Reply by KingOfTac** [reply]
Yes

**Reply by PoojaC94** [reply]
Thanks a lot!! I got it working!

**Reply by PoojaC94** [reply]
> yes. the flow is, click componentC->a list of child elements appear->click on one of them->onClick() is called-> check if one condition is satisfied->set isComponentOpen to true->componentB appears

@KingOfTac the other components have different functions such as opening an app link or opening a link in a different window. When I use $emit(), the toggleComponentB() works but the other functionalities don't.(even when I debug, I see the code flow working but it doesn't seem to open any links. However, if I provide property binding (`:onClick`) instead of @click (i.e. direct reference to onLinkClick() method), I could get to open links in a separate window or in-app(the toggleComponent doesn't work in this case)

Is there a way to get both the approaches working?

**Reply by KingOfTac** [reply]
@PoojaC94 can you create some code examples of what you're trying to do in a stackblitz or code sandbox. It will help me be able to help you a lot better if I can see what you're trying to do.
